dsl_parser/tasks.py

```diff
@@ -155,6 +155,100 @@ def _find_idd_creating_functions(plan) -> List[functions.IDDSpec]:
     return handler.dependencies
 
 
+def _find_rel_target_by_iddspec(relationships, idd_spec):
+    """Find the target instance that matches the relationship in IDDSpec
+
+    The node-instance can have many relationships, so we need to find one
+    that matches the target node name & type.
+    If there's multiple relationships of the same type to the same node,
+    then we will just select the first instance, because we have no way
+    to distinguish them.
+    (that case shouldn't be possible anyway)
+    """
+    assert idd_spec.scope == 'node_template_relationship'
+
+    target_name = idd_spec.context['target']  # target _node_ name
+    rel_type = idd_spec.context['type']
+
+    for relationship in relationships:
+        if (
+            target_name == relationship['target_name']
+            and rel_type == relationship['type']
+        ):
+            return relationship['target_id']
+
+    node_name = idd_spec.context['node']
+    raise ValueError(
+        'IDDs: relationship not found: {0} {1} {2}'
+        .format(node_name, rel_type, target_name),
+    )
+
+
+def _format_idds(plan, dep_specs):
+    """Format IDDSpecs into actual dependencies by choosing the node-instances.
+
+    IDDSpecs will contain just node name, but there can be many node instances
+    for a given node name, so we need to expand every such IDDSpec into
+    multiple actual IDD dicts.
+
+    This returns a list of dicts containing the keys:
+        - function_identifier - the IDDCreatingFunction path
+        - target_deployment - the deployment id, or a dict representing
+          a function call
+        - context - a dict containing possibly the keys SELF, TARGET, SOURCE -
+          this is the context that can be passed into function evaluation,
+          when evaluating the target_deployment (if it is a function)
+    """
+    idds = []
+    nis_by_node = {}
+    for ni in plan.get('node_instances', []):
```

Comment:
Not worth doing as part of this, but I wonder if in general it would make sense to make the plan generate node instances as a dict with lists of node instances keyed on node_id?

Reply:
(because I suspect we need to do something like this in several places?)

Reply:
OK- let's put it on the "one day, maybe today- maybe not" pile.